python/ray/air/preprocessors/stateful_preprocessor.py

```diff
@@ -0,0 +1,120 @@
+from typing import Callable, TYPE_CHECKING, Dict
+
+from ray.air.preprocessor import Preprocessor
+from ray.data import Dataset
+
+if TYPE_CHECKING:
+    import pandas
+
+
+class StatefulPreprocessor(Preprocessor):
+    """Implements a stateful preprocessor that fits on a Dataset.
+
+    This is meant to be generic and can be used to perform arbitrary stateful
+    preprocessing that cannot already be done through existing preprocessors.
+    Logic must be defined to perform fitting on a Ray Dataset and transforming
+    Pandas DataFrames.
+
+    Example:
+
+    .. code-block:: python
+
+        from typing import Dict
```

Comment:
If you do not define the types of keys and values, you can just not import `from typing import Dict` and use lower-case `dict` in the type hint.